### ISSUE TITLE
Complex.Log(Complex value, Complex baseValue)

### ISSUE THREAD

**Issue by Happypig375** [dotnet/runtime#40897]
## Background and Motivation

Currently, there are two overloads for Complex.Log, namely Complex.Log(Complex value) and Complex.Log(Complex value, double baseValue).

From https://en.wikipedia.org/wiki/Complex_logarithm#Logarithms_to_other_bases, the baseValue can be Complex and the current implementation of
```cs
        public static Complex Log(Complex value, double baseValue)
        {
            return Log(value) / Log(baseValue);
        }
```
will work just fine for this new overload.

It is unclear why baseValue is restricted to double as Complex will work just fine.

For usability, a Complex.Log(Complex value, Complex baseValue) overload should be added with an identical implementation as above.

## Proposed API

```diff
 namespace System.Numerics
 {
     public struct Complex
     {
+        public static Complex Log(Complex value, Complex baseValue);
     }
 }
```

## Usage Examples

```cs
var two = Complex.Log(new Complex(-5, 12), new Complex(2, 3)); // new Complex(2, 0)
```

## Alternative Designs

e.g. Python: `cmath.log(-5+12j, 2+3j)` is 2
## Risks

Nope.

**Comment by charlesroddie** [comment]
log(x) where x is complex: an arbitrary choice from a set `S={y:exp(y)=x}={k+2𝜋i n}`. This is a relatively simple set and the choice is understandable in terms of a branch cut. The answer will be some x for which exp(y)=x.

log_b(x) when b is real is very niche and probaly shouldn't have been part of the API, but doesn't add conceptual difficulties. It's just the set being chosen from is of the form {k + 2𝜋in/log b}, where log b is well defined as a real number. These are the solutions to b^y := exp(x * log b) = x.

log_b(x) when is ultra-niche: what is an application? It is also conceptually much more fiddly. The set being chosen from is {(k + 2𝜋i n)/(l + 2𝜋i m)}, i.e. an irregular grid on a 2 dimensional surface with dense points at rational numbers. I don't believe you can reconstruct this set from just one element, unlike the sets above. Any element of this set is a solution of b^y = x for some interpretation (i.e. branch cut) of b^.

So: it's good that this method is not there.

**Comment by WhiteBlackGoose** [comment]
@charlesroddie yes, unlike real-based logarithms, here you won't be able to change the branch. But as it's numerical, in most cases you really need *any* solution, exactly the same way as other multivalued functions work (say, square root or arc functions). Also, some other languages compute it the same way @Happypig375 proposed.

**Comment by charlesroddie** [comment]
@WhiteBlackGoose what do you mean by "any solution". It's not exactly the same as other multivalued solutions as explained above: since the pseudo-equation `b^y = x` has many interpretations depending on which branch you use to interpret `b^y`.

In what circumstance would you need any solution to any of the interpretations of `b^y=x`?

**Comment by tannergooding** [comment]
I'm going to leave this as an `api-suggestion` for now as there isn't a lot of demand and there are some initial concerns about drawbacks of exposing it.

Wikipedia explains one of the known caveats: https://en.wikipedia.org/wiki/Complex_logarithm#Logarithms_to_other_bases which would need to be addressed as part of the API design (that is, the proposal should call out the behavior, etc and use other languages such as C/C++ and Python as examples on what they do).
